filestore/core.py

```diff
@@ -73,16 +111,109 @@ def insert_datum(col, resource, datum_id, datum_kwargs, known_spec):
 
 
 def insert_resource(col, spec, resource_path, resource_kwargs,
-                    known_spec):
+                    known_spec, root):
     resource_kwargs = dict(resource_kwargs)
     if spec in known_spec:
         js_validate(resource_kwargs, known_spec[spec]['resource'])
 
     resource_object = dict(spec=str(spec),
                            resource_path=str(resource_path),
-                           resource_kwargs=resource_kwargs)
+                           root=str(root),
+                           resource_kwargs=resource_kwargs,
+                           uid=str(uuid.uuid4()))
 
     col.insert_one(resource_object)
-    # rename to play nice with ME
-    resource_object['id'] = resource_object.pop('_id')
-    return resource_object
+    # maintain back compatibility
+    resource_object['id'] = resource_object['uid']
+    resource_object.pop('_id')
+    return Document('resource', resource_object)
+
+
+def update_resource(update_col, resource_col, old, new, cmd, cmd_kwargs):
+    '''Update a resource document
+
+    Parameters
+    ----------
+    update_col : Collection
+        The collection to record audit trail in
+    resource_col : Collection
+        The resource collection
+
+    old : dict
+        The old resource document
+
+    new : dict
+        The new resource document
+
+    cmd : str
+        The name of the operation which generated this update
+
+    cmd_kwargs : dict
+        The arguments that went into the update (excluding the resource id)
+
+
+    Returns
+    -------
+    ret : Document
+        The new resource document
+
+    log_object : dict
+        The history object inserted (with oid removed)
+    '''
+    if old['uid'] != new['uid']:
+        raise RuntimeError('must not change the resource id')
+    uid = old['uid']
+    log_object = {'resource': uid,
+                  'old': old,
+                  'new': new,
+                  'time': ttime.time(),
+                  'cmd': cmd,
+                  'cmd_kwargs': cmd_kwargs}
+
+    update_col.insert_one(log_object)
+    result = resource_col.replace_one({'uid': uid}, new)
+    ret = resource_given_uid(resource_col, uid)
+    # TODO look inside of result
+    del result
+    log_object.pop('_id')
+    return ret, log_object
+
+
+def get_resource_history(col, resource):
+    uid = doc_or_uid_to_uid(resource)
+    cursor = col.find({'resource': uid}).sort('time')
+    for doc in cursor:
+        for k in ['new', 'old']:
+            d = doc[k]
+            d.pop('_id', None)
+            d['id'] = d['uid']
```

Comment:
this gets both (for now) there was comment about leaving the `id` field in there to preserve compatibility with the way that mongoengine returned things (which renames `_id` to `id`).  An early version of fs returned the ObjectID directly, then we swtched to returning the documents from ME, then to Doct.  I have a rough memory of there being atleast one pass at the ophyd level which would extract `['id']` from the return of `insert_resource` to maintain it's internal behavior. 

This line is here so that you get back exactly what you get back from `insert_resource`.